doc/design/xgboost_feature_column.md

```diff
@@ -42,38 +42,75 @@ The training process of XGBoost model inside SQLFlow are as follows:
 
 - Step 1: Read data from database. Call the `db.db_generator()` method to return a Python generator which yields each row in database. 
 - Step 2: Dump SVM file. Call the `dump_dmatrix()` method to write the raw data into an SVM file. This file is ready to be loaded as XGBoost DMatrix.
-- Step 3: Training. Load the dumped file to be XGBoost DMatrix and start to train.
+- Step 3: Training. Load the dumped file to be XGBoost DMatrix and start to train. The training process is performed by calling `xgboost.train` APIs.
 
-Between step 1 and 2, we can do some pre-processing to the raw data from database. In Python side, we can implement the feature column interfaces like:
+As discussed in [COLUMN clause for XGBoost](https://github.com/sql-machine-learning/sqlflow/issues/2190), there are 3 candidate ways to support feature column in XGBoost models:
+
+- Method 1. Perform feature column transformation during step 1 and step 2. Data pre-processing can be done before dumping to 
+SVM file. This method is suitable for offline training (both standalone and distributed), prediction and evaluation, since same
+transformation Python codes can be generated in both training, prediction and evaluation. But it is not suitable for online
+serving, because online serving usually uses other libraries or languages (like C++/Java), which does not support the
+transformation codes we generate in SQLFlow Python codes.
+
+- Method 2. Modify the training iteration of XGBoost, and insert transformation codes during each iteration. But it is not
+easy to modify the training iteration of XGBoost. Moreover, it is also not suitable for online serving for the same reason 
+as method 1.
+
+- Method 3. Combine data pre-processing and model training as a sklearn pipeline. Since sklearn pipeline can be saved as 
```

Comment:
I'm thinking whether we can try to develop a PMML post-processing tool. It will update the PMML file from model and add the transformation logic into it.

Reply:
Of course, `sklearn2pmml` also extends [many data pre-processing transformers](https://github.com/jpmml/jpmml-sklearn). But it may be another project or repo to develop a PMML post-preprocessing tool.